### PR TITLE
Date data type

### LINKED ISSUE #1076: Add support for `date` data type
### Search terms you've used
date, data type, addDate, getDate, setDate, removeDate

### Feature suggestion

There currently are `datetime` functions in solid-client, but none to specify a date without a time.

### Expected functionality/enhancement

Functions `getDate`, `addDate`, `setDate` and `removeDate` that are like their `datetime` equivalents, but disregard the time, and store the value as an `xsd:date`.

### Actual functionality/enhancement

Currently you're forced to create a `datetime`.

### Use Cases

Some times there's no time that would be relevant - e.g. for all-day events, like stating when Easter is this year.

### Additional information

To implement this:

- Add `xsd:date` to the `schemaTypes` we support [in `datatypes.ts`](https://github.com/inrupt/solid-client-js/blob/2ea2ab5b447bae24e5913b32520f6c960bdbac62/src/datatypes.ts#L34).
- Add `serializeDate` and `deserializeDate` to `datatypes.ts` in a way similar to [its datetime equivalents](https://github.com/inrupt/solid-client-js/blob/2ea2ab5b447bae24e5913b32520f6c960bdbac62/src/datatypes.ts#L69-L154), based on the spec that determines [what the stringified format of an `xsd:date` looks like](https://www.w3.org/TR/xmlschema-2/#date-lexical-representation).
- Add the relevant functions to `src/thing/get.ts`, `add.ts`, `set.ts` and `remove.ts` similar to the datetime equivalents.
- Don't forget to update the tests as well :)

Also see #1077 for a similar issue that you can pick up after this.

### PR DESCRIPTION
**Pull request by dgbarclay**
# Get/Set/Add/Remove methods for date data type

Fixes #1076.

# Checklist

- [x] All acceptance criteria are met.
- [] Relevant documentation, if any, has been written/updated.
- [x] The changelog has been updated, if applicable.
- [x] New functions/types have been exported in `index.ts`, if applicable.
- [x] New modules (i.e. new `.ts` files) are listed in the `exports` field in `package.json`, if applicable.
- [x] New modules (i.e. new `.ts` files) are listed in the `typedocOptions.entryPoints` field in `tsconfig.json`, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: src/datatypes.ts
Comment (Vinnl) [inline src/datatypes.ts:222]: I think this comment is irrelevant for `date` - since there's no time, a timezone offset can't push it into a different year. (i.e. in a timezone of `+02:00`, a time of `23:00` could mean the actual date could differ - but if there's no time, it can't be `23:00` either :)

File: src/datatypes.ts
Comment (Vinnl) [inline src/datatypes.ts:186]: Here, too, I think it's missing the fact that it can optionally also contain a timezone (i.e. `Z`, or `+01:30` or `-02:00`).

File: src/thing/add.test.ts
Comment (Vinnl) [inline src/thing/add.test.ts:612]:
```suggestion
  it("adds the given date value for the given predicate", () => {
```

File: src/thing/add.test.ts
Comment (Vinnl) [inline src/thing/add.test.ts:625]: I think this is the result of the `serializeDate` function just calling `toISOString()`, but I think this should be:

```suggestion
    ).toStrictEqual(["1990-11-12Z"]);
```

File: src/thing/get.test.ts
Comment (Vinnl) [inline src/thing/get.test.ts:1027]: There's a couple of other mentions of `datetime`s below too :)

```suggestion
  it("returns an empty array if no date values were found", () => {
```

File: src/thing/remove.test.ts
Comment (Vinnl) [inline src/thing/remove.test.ts:1110]: In this file too check for references to dateTime in your date tests :)

```suggestion
    const thingWithDate = getMockThingWithLiteralFor(
```

File: src/thing/set.ts
Comment (Vinnl) [inline src/thing/set.ts:111]:
```suggestion
 * Create a new Thing with existing values replaced by the given date for the given Property.
```

**Review by Vinnl — CHANGES_REQUESTED** [review]
A good start, but as far as I can see the serialise function does not match the spec, and the deserialize function is throwing away the timezone information. If you can fix that (and add the related tests) then I think we're probably good to go :)

File: src/datatypes.ts
Comment (dgbarclay) [reply src/datatypes.ts]: In this case, can deserialize essentially ignore the timezone and just return a Date object with T00:00:00?

Comment (Vinnl) [reply src/datatypes.ts]: Yeah I think that's the way to go. It's a bit fiddly though - I'd set it to 0:00 in the timezone set in the serialisation (or UTC, if none was set), but since JS `Date` objects only support either the local time or UTC, the time you get locally might still be somewhat different (e.g. it'd be 22:00 the previous day here).

Comment (Vinnl) [reply src/datatypes.ts]: Actually, maybe set it to 12:00? That way I think it should always match the local and UTC date regardless of the timezone, and the time part should be ignored anyway.

(Dates in JavaScript really is a mess. There's a proposal for a replacement, [Temporal](https://tc39.es/proposal-temporal/docs/), that looks like it will be added to JavaScript somewhere in the coming few years. But that doesn't help us now...

Comment (dgbarclay) [reply src/datatypes.ts]: Yeah, I was wondering what would happen for a date with a - timezone for setting at 0:00. I'll change now.

File: src/datatypes.ts
Comment (Vinnl) [inline src/datatypes.ts:214]: Ah right, we're basically stripping the timezone information from the input string, because there's no way to represent that using `Date`. Reeeaallly looking forward to JS getting a better date&time API :angry:

File: src/thing/remove.ts
Comment (Vinnl) [inline src/thing/remove.ts:191]: Hmm.... So this line gets me thinking: if the developer has to pass in a `Date` object not only with the correct date, but also with a certain time - how much value is it providing that we're using a `Date` object here?

Would it potentially make more sense to use a custom plain object here using only the properties we're interested in? So something like this:

```
type Date = {
  year: number;
  month: number;
  day: number;
  timezoneHourOffset?: number;
  timezoneMinuteOffset?: number;
};
```

Alternatively, if that's cumbersome too, maybe we should only check on `foundDate`'s and `value`'s year, month and day rather than adding 43200000ms?

Sorry, I'm being a little annoying here bring this up after implementing it one way, but it's the implementation experience that brought the insight...

**Review by Vinnl — COMMENTED** [review]
It mostly looks good, but I'm just being very annoying and having second thoughts about this approach - sorry! :/ I added a suggestion inline, but would love to hear your thoughts - and do speak up if you think it's rubbish, cause I'm certainly not sure about it.

As an aside, some process thingies to note that might be helpful in the future:

- If the PR is ready for another review, be sure to press the little "recycle" icon next to a reviewer's name, so they get a notification that they're expected to do something :)
image
- Before a PR can be merged, all checks should succeed.
image
You can click the "Details" link to see what went wrong - in this case, it looks like you didn't run Prettier on a number of files. This should generally run automatically when committing, so that's a bit weird. Either way, your editor probably can do it for you - in VSCode, you can find "Format Selection" if you press Ctrl+Shift+P.
- The PR template mentions "Commits in this PR are minimal and have descriptive commit messages." During review you can sometimes collect a few minor fixing commits, but please when merging squash them all together if the PR is essentially just a single change. (Ideally you'd rebase locally, creating the commits that make sense, but that does involve some more advanced Git knowledge :) ) Not relevant yet, but when finishing up the PR :)

File: src/thing/remove.ts
Comment (dgbarclay) [reply src/thing/remove.ts:191]: Yeah, it's a fair point. On the set/get/add end, the user can pass in any date object and the serializing will handle all of the annoying date-related things. So, for the user, it's still fairly simple/easy. For example, the following code will add "1990-11-12Z" to the thing, ignoring the time :

```
const updatedThing = addDate(     
 thing,      
"https://some.vocab/predicate", 
 new Date(Date.UTC(1990, 10, 12, 3, 3, 3))
);
```
The difficulty, as you say, is on the remove side. In this case, they would have to provide the full date object with the time set to 0 in order to get a match, not ideal or intuitive.

I'm reluctant to completely ditch the Date object because I like how smoothly it works for appending to things, but for the remove method, we need a better solution. In my opinion, I think modifying the code to check for a date match should be enough. In that use case, a user just has to match the date they set originally which I think is pretty okay from a UX point of view, and, we can remove the clunky 43200000ms...

Comment (dgbarclay) [reply src/thing/remove.ts:191]: Also, thanks for git tips, useful stuff to know for the future.

Comment (Vinnl) [reply src/thing/remove.ts:191]: OK, sounds good!

And how about time: would a custom object make sense there, since it's somewhat more clunky to construct a `Date` just for the time?

Comment (dgbarclay) [reply src/thing/remove.ts:191]: Yeah, I think that's a good idea. Returning the user 1970 Jan 1st attached to their time isn't nice. Something like this...

```
type Time = {
  hour: number;
  minute: number;
  second: number;
  timezoneHourOffset?: number;
  timezoneMinuteOffset?: number;
};
```

We'd need to document this change well, is that something I need to do before the PR is merged?

Comment (Vinnl) [reply src/thing/remove.ts:191]: You're still working on time right? So it needn't block this PR.

For the time PR/branch, you can define that type and then export it from `index.ts` and add a TypeDoc comment block of documentation to that type as well.

Comment (dgbarclay) [reply src/thing/remove.ts:191]: Yeah oops for the other PR. Okay, I'll do that.

**Review by Vinnl — APPROVED** [review]
Quite a few more edge cases here than I expected, but you did it! Thanks a lot :)

**Comment by Vinnl** [comment]
Ah, and of course there were parallel changes in `main` that this has to align with too: abeda783.

File: src/thing/remove.ts
Comment (dgbarclay) [reply src/thing/remove.ts]: Ahhh I see, thanks for sorting.

File: src/datatypes.test.ts
Comment (dgbarclay) [reply src/datatypes.test.ts]: That's a fun one, always have reviewers in different timezones!!

**Pull request merged**
